File: src/basic.erl
-module(basic).

-export([answer/0, greet/1, sum/1]).

answer() -> {ok, 42}.

greet(Name) when is_binary(Name) ->
    io:format(<<"Hello ", Name/binary, "\n">>);
greet(Name) when is_list(Name) ->
    io:format("Hello " ++ Name ++ "\n").

%% Sum numbers in a list
sum(List) ->
    sum(List, 0). % assume empty list = 0

sum([], Acc) ->
    Acc;
sum([N | Rest], Acc) ->
    sum(Rest, N + Acc).

File: lib/basic.ex
defmodule Basic do
  def answer, do: {:ok, 42}

  def greet(name)

  def greet(name) when is_binary(name) do
    IO.puts("Hello " <> name)
  end

  def greet(name) when is_list(name) do
    IO.puts('Hello' ++ name)
  end

  # Sum numbers in a list
  def sum(list, acc \\ 0)

  def sum([], acc) do
    acc
  end

  def sum([n | rest], acc) do
    sum(rest, n + acc)
  end
end
